File: src/main/cppclient/Client2.cpp
#include <iostream>
#include "Contrat.hh"
#include <omniORB4/CORBA.h>
#include <omniORB4/Naming.hh>

int main(int argc, char* argv[]) {
    try {
        // Initialisation de l'ORB (Object Request Broker)
        CORBA::ORB_ptr orb = CORBA::ORB_init(argc, argv);

        // Résolution de la référence initiale du service de nommage
        CORBA::Object_var obj = orb->resolve_initial_references("NameService");
        CosNaming::NamingContext_var namingContext = CosNaming::NamingContext::_narrow(obj.in());

        // Résolution du nom "Promotion" dans le service de nommage
        CosNaming::Name name(1);
        name.length(1);
        name[0].id = CORBA::string_dup("Promotion");
        CORBA::Object_var obj_ref = namingContext->resolve(name);
        
        // Récupération de l'objet "Promotion" via un narrowing
        Contrat::Promotion_var promotion = Contrat::Promotion::_narrow(obj_ref.in());

        // Ajout d'étudiants à la promotion
        promotion->ajouter_un_etudiant("A", "B", 123);
        Contrat::Etudiant_var etudiant1 = promotion->rechercher_un_etudiant(123);

        // Ajout d'emprunts de livres pour l'étudiant
        etudiant1->Emprunter_un_livre("Physique Quantique", "Feynman", "AB","1980");
        etudiant1->Emprunter_un_livre("Chimie", "Niels Bohr", "ABC","1905");


        // afficher livres emrptunes 
        std::cout << "Livres emprunt'es par etudiant 1 " << etudiant1->Liste_des_livres() << std::endl;

        
        // Destruction de l'ORB
        orb->destroy();
    } catch (const CORBA::Exception& ex) {
        // Gestion des exceptions CORBA
        std::cerr << "CORBA exception: " << ex._name() << std::endl;
    }
}

File: src/main/java/com/example/Client2.java
package com.example;
import com.example.*;
import com.example.Contrat.Etudiant;
import com.example.Contrat.EtudiantHelper;
import com.example.Contrat.Promotion;
import com.example.Contrat.PromotionHelper;

import org.omg.CORBA.ORB;
import org.omg.CosNaming.NamingContextExt;
import org.omg.CosNaming.NamingContextExtHelper;

import java.util.Arrays;

public class Client2 {
    public static void main(String[] args) {
        try {
            ORB orb = ORB.init(args, null); //init orb
            org.omg.CORBA.Object objRef = orb.resolve_initial_references("NameService"); // référence de l'objet depuis le serveur de nommage
            NamingContextExt ncRef = NamingContextExtHelper.narrow(objRef);
            Promotion promotionRef = PromotionHelper.narrow(ncRef.resolve_str("Promotion")); // obtenir la reference de l'objet promotion
            // Déclaration de plusieurs étudiants
            promotionRef.ajouter_un_etudiant("A", "B", 12345);
            promotionRef.ajouter_un_etudiant("B", "A", 54321);
//            // Recherche des étudiants et ajout des emprunts de livres
            Etudiant etudiant1 = promotionRef.rechercher_un_etudiant(12345);
            Etudiant etudiant2 = promotionRef.rechercher_un_etudiant(54321);
//
            etudiant1.Emprunter_un_livre("Orientalism",  "Edward Said", "Penguin","1970");
            etudiant1.Emprunter_un_livre("Incerto",  "Nassim Nicholas Taleb", "Penguin","2003");

            etudiant1.Emprunter_un_livre("Profit Over People",  "Chomsky", "Penguin","2000");
            etudiant1.Emprunter_un_livre("Bed of Procrustes",  "Nassim Nicholas Taleb", "Penguin","2008");
//             Afficher les livres empruntes  pour chaque étudiant
            System.out.println(" livres empruntes par etudiant 1");
            System.out.println(Arrays.toString(etudiant1.Liste_des_livres()));
            System.out.println(" livres empruntes par etudiant 2");
            System.out.println(Arrays.toString(etudiant2.Liste_des_livres()));



        } catch (Exception e) {
            System.err.println("Erreur : " + e.getMessage());
            e.printStackTrace();
        }
    }
}
